validator/enforce_labels.rego

```diff
@@ -35,35 +26,58 @@ deny[{
 	lib.get_constraint_params(constraint, params)
 	asset := input.asset
 
-	mandatory_labels := params.mandatory_labels[_]
-	resource_is_to_scan(asset)
-	not label_is_valid(mandatory_labels, asset)
+	mandatory_label := params.mandatory_labels[_]
+	label_value_pattern := mandatory_label[label_key]
+
+	tested_resource_types := [
+		"cloudresourcemanager.googleapis.com/Project",
+		"storage.googleapis.com/Bucket",
+		"compute.googleapis.com/Instance",
+		"compute.googleapis.com/Image",
+		"compute.googleapis.com/Disk",
+		"compute.googleapis.com/Snapshot",
+		"google.bigtable.Instance",
+		"sqladmin.googleapis.com/Instance",
+	]
+
+	non_standard_types := ["sqladmin.googleapis.com/Instance"]
 
-	message := sprintf("%v doesn't have a required label.", [asset.name])
+	resource_types_to_scan := lib.get_default(params, "resource_types_to_scan", tested_resource_types)
 
-	metadata := {"resource": asset.name, "mandatory_labels": mandatory_labels}
+	resource_is_to_scan(asset, resource_types_to_scan)
+
+	not label_is_valid(label_key, label_value_pattern, asset, non_standard_types)
+
+	message := sprintf("%v's label is in violation.", [asset.name])
+	metadata := {"resource": asset.name, "label_in_violation": label_key}
```

Comment:
does this mean we'll have one violation per missing/invalid label? do we want to generate one violation per asset regardless of how many labels are missing/invalid instead?

Reply:
You will indeed get one violation per missing/invalid label, which was the intent here. This is easier to troubleshoot for the end user, and in the context of multiple people working on the violations (for instance different teams working on the same assets), this makes it easier to split the tasks / fixes.